<center><b>Python Programming for Multilingual Text</b></center>
<center>2-4: Introduction to Text Analysis with Montagu Corpus</center>

---

# Introduction to Text Analysis with Montagu Corpus

## Overview

This notebook introduces some basic string methods using examples from Lady Mary Montagu's letters. It applied and thus reinforces what we have learnt in the previous notebooks about Python more generally.

1. Data Import: We will read in `Letter 35` and prepare it for analysis
2. String Operation: We will look at some string operations
3. Counting Words: Tokenization, Text normalization, most frequent words
4. Stopwords: stopword removal, NLTK
5. Sentences: NLTK, initial discussions of syntax

## Step 1: Letter 35

In [1]:
letter_35 = open("../../output/montagu_letters/letter_35.txt").read()
print(letter_35)

LET. XXXV.

To THE ABBOT ----.

_Constantinople, May_ 29. O. S.

I HAVE had the advantage of very fine weather, all my journey; and as
the summer is now in its beauty, I enjoyed the pleasure of fine
prospects; and the meadows being full of all sorts of garden flowers,
and sweet herbs, my berlin perfumed the air as it pressed them.  The
grand signior furnished us with thirty covered waggons for our
baggage, and five coaches of the country for my women.  We found the
road full of the great spahis and their equipages coming out of Asia
to the war.  They always travel with tents; but I chose to ly in
houses all the way.  I will not trouble you with the names of the
villages we passed, in which there was nothing remarkable, but at
Ciorlei, where there was a _conac_, or little seraglio, built for the
use of the grand signior, when he goes this road.  I had the
curiosity to view all the apartments destined for the ladies of his
court.  They were in the midst of a thick grove of trees, made fr

For this notebook, we want to only work with the body of the letter and not the metadata. 

<div style="background: rgba(167, 139, 250, 0.12); border-radius: 6px; padding: 0.6em 1em; margin: 1.5em 0; font-weight: 500;">
  Let's remove the metadata on the fly.
</div>



### 🥊 Challenge 1: Remove the metadata

Hint: there are probably more ways to tackle this than we could think of. This is the beauty of coding, we will each have a unique approach to solving this

In [4]:
## YOUR CODE HERE

This is how I would do it (in pseudocode)

- find the start of the letter
    - it begins with "I HAVE"
    - search for "I HAVE" in the letter
    - verify that there is only one instance of it
    - get the index using .find() method
- use the start index to split the metadata from the body
    - we can slice a string using indices
    - remember that the end index in a slice is *exclusive*
    - verify that the split is correct
- delete everything until the letter body
    - this will modify the variable 'letter_35'

In [3]:
# I know where the body of the letter begins

letter_start = letter_35.find("I HAVE")
print(letter_start)

65


In [6]:
# let's get the metadata to verify our approach
letter_metadata = letter_35[:letter_start]
print(letter_metadata)

LET. XXXV.

To THE ABBOT ----.

_Constantinople, May_ 29. O. S.




In [4]:
# Looking good! We can now use the same logic to modify our letter_35
# this is an irreversable change because we are assigning it back to the variable itself!

letter_35 = letter_35[letter_start:]
print(letter_35)

I HAVE had the advantage of very fine weather, all my journey; and as
the summer is now in its beauty, I enjoyed the pleasure of fine
prospects; and the meadows being full of all sorts of garden flowers,
and sweet herbs, my berlin perfumed the air as it pressed them.  The
grand signior furnished us with thirty covered waggons for our
baggage, and five coaches of the country for my women.  We found the
road full of the great spahis and their equipages coming out of Asia
to the war.  They always travel with tents; but I chose to ly in
houses all the way.  I will not trouble you with the names of the
villages we passed, in which there was nothing remarkable, but at
Ciorlei, where there was a _conac_, or little seraglio, built for the
use of the grand signior, when he goes this road.  I had the
curiosity to view all the apartments destined for the ladies of his
court.  They were in the midst of a thick grove of trees, made fresh
by fountains; but I was most surprised to see the walls almos

## Step 2: Simple String Operations

First, let's gather some basic yet important statistics about our letter

In [8]:
print(f"Length of letter 35: {len(letter_35)} characters")
print(f"Number of words in letter 35: {len(letter_35.split())} words")

Length of letter 35: 9777 characters
Number of words in letter 35: 1799 words


This seems like a long letter, almost 1800 words!

<div style="background: rgba(167, 139, 250, 0.12); border-radius: 6px; padding: 0.6em 1em; margin: 1.5em 0; font-weight: 500;">
  Let's see what else is in the letter other than words. Perhaps punctuation? How many punctuation marks are in this letter?
</div>


In [5]:
import string

[String](https://www.w3schools.com/python/ref_module_string.asp) is a Python module, that is built into the language itself. It provides constants and classes for common string operations.

We use it to access predefined sets of characters (letters, digits, punctuation)

In [13]:
print(type(string.punctuation))
print(string.punctuation)

<class 'str'>
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


### 🥊 Challenge 2: Count Punctuation

Hint: remember what we learnt about for-loops in notebook `2-2_iteration`


In [ ]:
# Take a second and try to implement a solution

## YOUR CODE HERE

Here is how I would do it

- string.punctuation gives us all ASCII punctuation characters
- start a count at 0
- loop over each character in my letter and see if it is punctuation, meaning included in string.punctuation
- if it is increase the count by 1

In [14]:
# initiate count
punctuation_count = 0

# loop over letter_35
for char in letter_35:
    if char in string.punctuation:
        punctuation_count += 1

print(punctuation_count)

295


Another way to approach this is to loop over each character in string.punctuation and for each character call the count function on letter_35 to get how many of a given character appears in letter_35. count() returns an integer, which we can then add to our count

In [15]:
# initiate count
punct_count = 0

# loop over string.punctuation
for char in string.punctuation:
    # count how many of a given punctuation character appears in letter_35
    punct_count += letter_35.count(char)

print(punct_count)

295


The result is the same in both approaches! Which do you prefer?

Let's see what percentage of our letter was punctuation

In [16]:
print("Percentage of punctuation marks in letter 35:")
print(f"{(punctuation_count / len(letter_35)) * 100:.2f}%")

Percentage of punctuation marks in letter 35:
3.02%


### 🥊 Challenge 3: Unique Punctuation Marks

Hint: Remember notebook `1-6_data_structures`?

<div style="background: rgba(167, 139, 250, 0.12); border-radius: 6px; padding: 0.6em 1em; margin: 1.5em 0; font-weight: 500;">
  Which punctuation marks appear in this letter?
</div>

In [ ]:
### YOUR CODE HERE

Here is how I would do it:

- create a list
- loop over string.punctuation
- check if punctuation mark is in letter_35
- if yes, add punctuation mark to the list

In [18]:
# empty list

punctuation_marks = []

for punct in string.punctuation:
    if punct in letter_35:
        punctuation_marks.append(punct)
print(punctuation_marks)

['&', "'", '(', ')', ',', '-', '.', ':', ';', '_']


### 🥊 Challenge 4: How Many of Each Punctuation

Hint: Remember dictionaries from notebook `1-6_data_structures`?

<div style="background: rgba(167, 139, 250, 0.12); border-radius: 6px; padding: 0.6em 1em; margin: 1.5em 0; font-weight: 500;">
  Count how many times each punctuation mark appears in the text
</div>

In [ ]:
### YOUR CODE HERE

Here is how I would do it:

- initiate an empty dictionary
- loop over each punctuation mark in string.punctuation
- use count() to see how many times they appear in letter_35
- using the punctuation mark as the key and the count as the value, add it to the dictionary

In [6]:
# initiate dictionary
punct_amount = {}

# loop over punctuation
for char in string.punctuation:
    # get a count
    count = letter_35.count(char)
    # if count exists
    if count:
        # add the punctuation and its count to the dictionary
        punct_amount[char] = count

print(punct_amount)


{'&': 2, "'": 11, '(': 4, ')': 4, ',': 158, '-': 8, '.': 70, ':': 2, ';': 30, '_': 6}


Alternatively, we have an amazing Python module to help with this problem: [collections](https://docs.python.org/3/library/collections.html) and especially `Counter`

In [7]:
from collections import Counter

Here is the trick, Counter() works the best for tallying over a list of strings, which as we will see later, is an amazing feature. 

It also means that we need to first filter our letter_35, keeping only the punctuation marks that appear in the letter.

In [9]:
# we are using list comprehension here, which is the other iteration method we studied in notebook 2-2_iterations

punct_only_letter = [char for char in letter_35 if char in string.punctuation]

print(punct_only_letter)

[',', ';', ',', ';', ',', ',', '.', ',', '.', '.', ';', '.', ',', ',', ',', '_', '_', ',', ',', ',', '.', '.', ',', ';', ',', '.', ',', ';', ',', '.', ':', '_', ';', ',', ';', ',', "'", '.', '_', ',', ',', '.', ',', '.', '-', ',', ',', '-', '.', '.', ',', ';', ',', '.', '-', ',', ',', ',', '-', '.', "'", ',', ';', ',', ',', ';', ',', ',', '(', ')', ',', '.', ',', ',', ',', '.', ';', ',', '.', ',', ',', ',', '.', ',', ';', ',', ',', ';', ',', ';', ',', ',', '.', ',', ',', ';', ',', '.', "'", '_', '_', ',', '.', ',', ',', ',', ',', ',', '.', ',', ';', ',', ',', ',', '.', '.', ';', ',', ',', ',', ',', ',', ',', '.', ',', ',', '.', '.', ',', ',', ',', ';', ',', ',', '.', ',', '.', '-', '-', '-', '-', "'", ',', ';', "'", '.', '.', "'", '.', ',', ',', ',', ',', '.', '.', ',', '.', ',', ',', ';', ',', ',', ',', "'", ';', ',', '.', ',', ',', '.', ',', '.', ',', ',', '.', ',', ',', ',', ',', ';', "'", ';', ',', ',', ',', ',', ';', ',', '.', '.', '.', ',', ',', ';', ',', ',', ',', '.', ',', ',',

In [8]:
punct_cnt = Counter(punct_only_letter)

print(punct_cnt)

Counter({',': 158, '.': 70, ';': 30, "'": 11, '-': 8, '_': 6, '(': 4, ')': 4, ':': 2, '&': 2})


This might seem like an overly complicated way to just count punctuation marks but Counter will be super useful in the next step

<mark style="background: rgba(167, 139, 250, 0.3); color: inherit; border-radius: 3px; padding: 0 2px;">There are so many other things we can count, whitespace, numbers, capitalized letters, etc. We strongly urge you to implement your own approaches for dissecting and counting different aspects of the letter</mark>


## Step 3: Counting Words

<div style="background: rgba(167, 139, 250, 0.12); border-radius: 6px; padding: 0.6em 1em; margin: 1.5em 0; font-weight: 500;">
  To count words, we need to find words.
</div>


Now that we warmed up by counting punctuation, let's tackle the elephant in the room: **tokens**

You probably heard of tokens in the context of large language models, and we will get to that in the next few days, but for the purposes of this notebook, let's stick to an older definition of a token:

> Tokenization is the process of splitting text into _something_ - often words. **Each word is called a "token"** and a word such as "the" might adhere to multiple tokens of "the" within a text based on its capitalization, punctuation, etc.

The `.split()` method allows us to split the text based on some sort of separator. The default is blank and will split on the white spaces between words. As such, we are using the `.split()` method as our **tokenizer**. 

In [10]:
# Let's split our letter into words

letter_35.split()

['I',
 'HAVE',
 'had',
 'the',
 'advantage',
 'of',
 'very',
 'fine',
 'weather,',
 'all',
 'my',
 'journey;',
 'and',
 'as',
 'the',
 'summer',
 'is',
 'now',
 'in',
 'its',
 'beauty,',
 'I',
 'enjoyed',
 'the',
 'pleasure',
 'of',
 'fine',
 'prospects;',
 'and',
 'the',
 'meadows',
 'being',
 'full',
 'of',
 'all',
 'sorts',
 'of',
 'garden',
 'flowers,',
 'and',
 'sweet',
 'herbs,',
 'my',
 'berlin',
 'perfumed',
 'the',
 'air',
 'as',
 'it',
 'pressed',
 'them.',
 'The',
 'grand',
 'signior',
 'furnished',
 'us',
 'with',
 'thirty',
 'covered',
 'waggons',
 'for',
 'our',
 'baggage,',
 'and',
 'five',
 'coaches',
 'of',
 'the',
 'country',
 'for',
 'my',
 'women.',
 'We',
 'found',
 'the',
 'road',
 'full',
 'of',
 'the',
 'great',
 'spahis',
 'and',
 'their',
 'equipages',
 'coming',
 'out',
 'of',
 'Asia',
 'to',
 'the',
 'war.',
 'They',
 'always',
 'travel',
 'with',
 'tents;',
 'but',
 'I',
 'chose',
 'to',
 'ly',
 'in',
 'houses',
 'all',
 'the',
 'way.',
 'I',
 'will',
 'not

In [11]:
# looking good! let's actually assign it to a variable

words = letter_35.split()

Not the best variable name, but it does the job!

Let's investigate further

In [19]:
# let's look at this one common word:
words[50]

'them.'

See, .split() did not remove punctuation, which means:

In [20]:
words.index("them")

205

In [21]:
words.index("them,")

1186

In [25]:
words.count("them")

7

Since our goals is to count words, we need to normalize `them` a little. We have several different versions of essentially the same word.

Text normalization is the process of transforming text into a consistent, standardized form so that all forms of the word are all counted as the same word. Common steps include *lowercasing*, *removing punctuation*, and *stripping extra whitespace*.

### 🥊 Challenge 5: Remove Punctuation

Hint: We just did something very similar!

In [ ]:
### YOUR CODE HERE

Here is how I would remove punctuation:

`LOOK AT MY CODE AND TELL ME HOW I DID IT`


In [22]:
words_clean = [word.strip(string.punctuation) for word in words]

In [23]:
words_clean.index("them,")

ValueError: list.index(x): x not in list

In [24]:
words_clean.count("them")

14

### 🥊 Challenge 6: Lowercase everything

While we are at it, let's lowercase everything too


In [ ]:
##  YOUR CODE HERE

Here is how I would remove punctuation:

`LOOK AT MY CODE AND TELL ME HOW I DID IT`

In [26]:
words_extra_clean = []

for word in words_clean:
    words_extra_clean.append(word.lower())

print(words_extra_clean[:50])

['i', 'have', 'had', 'the', 'advantage', 'of', 'very', 'fine', 'weather', 'all', 'my', 'journey', 'and', 'as', 'the', 'summer', 'is', 'now', 'in', 'its', 'beauty', 'i', 'enjoyed', 'the', 'pleasure', 'of', 'fine', 'prospects', 'and', 'the', 'meadows', 'being', 'full', 'of', 'all', 'sorts', 'of', 'garden', 'flowers', 'and', 'sweet', 'herbs', 'my', 'berlin', 'perfumed', 'the', 'air', 'as', 'it', 'pressed']


In [27]:
# Let's compare the counts for a very common word: the

print("Number of 'the' in words_clean:", words_clean.count("the"))

print("Number of 'the' in words_EXTRA_clean:", words_extra_clean.count("the"))


Number of 'the' in words_clean: 91
Number of 'the' in words_EXTRA_clean: 100


<mark style="background: rgba(167, 139, 250, 0.3); color: inherit; border-radius: 3px; padding: 0 2px;">Now, we can continue with our slightly humourously titled words_extra_clean list or we can actually apply the transformations that we are happy with back to our original list, words.</mark>

In [29]:
# remember that we can chain functions!

words = [word.strip(string.punctuation).lower() for word in words]

words[:50]


['i',
 'have',
 'had',
 'the',
 'advantage',
 'of',
 'very',
 'fine',
 'weather',
 'all',
 'my',
 'journey',
 'and',
 'as',
 'the',
 'summer',
 'is',
 'now',
 'in',
 'its',
 'beauty',
 'i',
 'enjoyed',
 'the',
 'pleasure',
 'of',
 'fine',
 'prospects',
 'and',
 'the',
 'meadows',
 'being',
 'full',
 'of',
 'all',
 'sorts',
 'of',
 'garden',
 'flowers',
 'and',
 'sweet',
 'herbs',
 'my',
 'berlin',
 'perfumed',
 'the',
 'air',
 'as',
 'it',
 'pressed']

### 🥊 Challenge 7: Find unique words

In [ ]:
## YOUR CODE HERE

In [39]:
# Remember sets?

unique_words = set(words)

unique_words

{'a',
 'about',
 'acceptable',
 'admitted',
 'advantage',
 'afraid',
 'after',
 'again',
 'agreeable',
 'air',
 'aleppo',
 'all',
 'almost',
 'already',
 'always',
 'am',
 'ambassadors',
 'an',
 'ancient',
 'anciently',
 'and',
 'another',
 'antiquaries',
 'antiquity',
 'any',
 'apartment',
 'apartments',
 'appears',
 'appointed',
 'arabia',
 'arches',
 'are',
 'arms',
 'arrived',
 'as',
 'asia',
 'asked',
 'assured',
 'at',
 'augustine',
 'author',
 'back',
 'baggage',
 'be',
 'bears',
 'beauties',
 'beautiful',
 'beauty',
 'bed',
 'been',
 'befel',
 'before',
 'being',
 'belief',
 'believe',
 'believed',
 'belonging',
 'beloved',
 'berlin',
 'besides',
 'bespoke',
 'best',
 'big',
 'bigger',
 'bishop',
 'bliss',
 'body',
 'boys',
 'branch',
 'bridge',
 'bringing',
 'built',
 'bujuk',
 'burning',
 'burying',
 'but',
 'by',
 'c',
 'cairo',
 'called',
 'calling',
 'can',
 'candle',
 'cannot',
 'care',
 'carved',
 'cekmege',
 'celestial',
 'certain',
 'certainly',
 'children',
 'chose',


In [41]:
len(unique_words)

681

### 🥊 Challenge 8: Count unique words



In [ ]:
### YOUR CODE HERE

In [31]:
# Remember Counter()

unique_word_counts = Counter(words)

unique_word_counts

Counter({'the': 100,
         'of': 88,
         'a': 61,
         'to': 49,
         'i': 45,
         'and': 40,
         'in': 30,
         'is': 23,
         'it': 22,
         'that': 21,
         'are': 21,
         'have': 20,
         'they': 20,
         'with': 18,
         'for': 18,
         'their': 17,
         'very': 15,
         'all': 15,
         'but': 15,
         'them': 14,
         'my': 12,
         'not': 12,
         'which': 12,
         'was': 12,
         'me': 12,
         'who': 12,
         'this': 11,
         'as': 10,
         'at': 10,
         'when': 9,
         'his': 9,
         'one': 9,
         'some': 9,
         'we': 8,
         'you': 8,
         'her': 8,
         'or': 7,
         'little': 7,
         'by': 7,
         'on': 7,
         'so': 7,
         'than': 7,
         'those': 7,
         'great': 6,
         'no': 6,
         'from': 6,
         'tis': 6,
         'can': 6,
         'other': 6,
         'be': 6,
         'had': 

Another neat piece of code is .most_common()

Gotta love the Counter tools

In [32]:
unique_word_counts.most_common(10)

[('the', 100),
 ('of', 88),
 ('a', 61),
 ('to', 49),
 ('i', 45),
 ('and', 40),
 ('in', 30),
 ('is', 23),
 ('it', 22),
 ('that', 21)]

<div style="background: rgba(167, 139, 250, 0.12); border-radius: 6px; padding: 0.6em 1em; margin: 1.5em 0; font-weight: 500;">
 Does something stand out to you about the top 10 most frequent words?
</div>

## Step 4: Stopwords

[Stop words](https://en.wikipedia.org/wiki/Stop_words) are the most common words in a language, and may or may not add information about the content of the analysis.

There are a lot of ways to handle stopwords in a corpus. If your (English language) texts have over a million or so words, you can simply get rid of top 100-200 most frequent words, depending on the specifics of your corpus.

Otherwise, another very common method is to use lists of stopwords.

Similar to what we did with punctuation, we will use an external library to acquire a list of stopword, specificially designed for English

In [34]:
import nltk
nltk.download("stopwords")
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/mervetekgurler/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


[NLTK](https://www.nltk.org/) stands for Natural Language Toolkit and is a Python library that provides tools for working with human language data (over 50 languages!), such as tokenizing text, removing stopwords, and analyzing word frequencies.

We will talk more about NLTK! For now, let's just take a quick look at their stopwords

In [35]:
stopwords_en = stopwords.words("english")
print(stopwords_en)

['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', "aren't", 'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', 'couldn', "couldn't", 'd', 'did', 'didn', "didn't", 'do', 'does', 'doesn', "doesn't", 'doing', 'don', "don't", 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'hadn', "hadn't", 'has', 'hasn', "hasn't", 'have', 'haven', "haven't", 'having', 'he', "he'd", "he'll", 'her', 'here', 'hers', 'herself', "he's", 'him', 'himself', 'his', 'how', 'i', "i'd", 'if', "i'll", "i'm", 'in', 'into', 'is', 'isn', "isn't", 'it', "it'd", "it'll", "it's", 'its', 'itself', "i've", 'just', 'll', 'm', 'ma', 'me', 'mightn', "mightn't", 'more', 'most', 'mustn', "mustn't", 'my', 'myself', 'needn', "needn't", 'no', 'nor', 'not', 'now', 'o', 'of', 'off', 'on', 'once', 'only', 'or', 'other', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 're', 's', 'same', 'shan', "shan't", 'she

In [36]:
words_no_stop = []

for word in words:
    if word not in stopwords.words('english'):
        words_no_stop.append(word)

print(words_no_stop)

['advantage', 'fine', 'weather', 'journey', 'summer', 'beauty', 'enjoyed', 'pleasure', 'fine', 'prospects', 'meadows', 'full', 'sorts', 'garden', 'flowers', 'sweet', 'herbs', 'berlin', 'perfumed', 'air', 'pressed', 'grand', 'signior', 'furnished', 'us', 'thirty', 'covered', 'waggons', 'baggage', 'five', 'coaches', 'country', 'women', 'found', 'road', 'full', 'great', 'spahis', 'equipages', 'coming', 'asia', 'war', 'always', 'travel', 'tents', 'chose', 'ly', 'houses', 'way', 'trouble', 'names', 'villages', 'passed', 'nothing', 'remarkable', 'ciorlei', 'conac', 'little', 'seraglio', 'built', 'use', 'grand', 'signior', 'goes', 'road', 'curiosity', 'view', 'apartments', 'destined', 'ladies', 'court', 'midst', 'thick', 'grove', 'trees', 'made', 'fresh', 'fountains', 'surprised', 'see', 'walls', 'almost', 'covered', 'little', 'distiches', 'turkish', 'verse', 'wrote', 'pencils', 'made', 'interpreter', 'explain', 'found', 'several', 'well', 'turned', 'though', 'easily', 'believed', 'lost', 'mu

In [38]:
words_no_stop_counts = Counter(words_no_stop)
words_no_stop_counts.most_common(10)

[('one', 9),
 ('little', 7),
 ('great', 6),
 ('tis', 6),
 ('fine', 5),
 ('night', 5),
 ('good', 5),
 ('greek', 5),
 ('upon', 5),
 ('women', 4)]

### 🥊 Challenge 9: How many unique words do we have after stopword removal?


In [40]:
# remember how we can nest functions?

print(len(set(words_no_stop)))

589


From 681 to 589! We removed about 13.5% of our unique words

## Step 5: Sentences

Another common way to approach text analysis is to start with sentences. If tokenization assumes that the smallest unit of analysis is a word, splitting text into sentences, assumes that the smallest unit of analysis is a sentence. 

Splitting into sentences is super useful for many syntax analysis. Sentences retain a lot more context than individual words. Let's look at this sentence from Letter 35

`The burying fields about it are certainly much larger than the whole city.`

`The burying fields` here refers to cemeteries, whereas in another context `burying` could be a verb. When we split the whole text into words, we inevitably flatten some of the meaning in these texts.

Luckily, there is another NLTK module for sentence tokenization

In [43]:
nltk.download('punkt')
nltk.download('punkt_tab')
from nltk.tokenize import sent_tokenize

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/mervetekgurler/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/mervetekgurler/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [46]:
sentences = sent_tokenize(letter_35)

In [47]:
# Here is another way to do a fancy print statement

print(f"Number of sentences in letter 35: {len(sentences)}")

Number of sentences in letter 35: 68


In [48]:
# And another wat

counter = 1
for sentence in sentences:
    print(counter) 
    print(sentence)
    counter += 1

1
I HAVE had the advantage of very fine weather, all my journey; and as
the summer is now in its beauty, I enjoyed the pleasure of fine
prospects; and the meadows being full of all sorts of garden flowers,
and sweet herbs, my berlin perfumed the air as it pressed them.
2
The
grand signior furnished us with thirty covered waggons for our
baggage, and five coaches of the country for my women.
3
We found the
road full of the great spahis and their equipages coming out of Asia
to the war.
4
They always travel with tents; but I chose to ly in
houses all the way.
5
I will not trouble you with the names of the
villages we passed, in which there was nothing remarkable, but at
Ciorlei, where there was a _conac_, or little seraglio, built for the
use of the grand signior, when he goes this road.
6
I had the
curiosity to view all the apartments destined for the ladies of his
court.
7
They were in the midst of a thick grove of trees, made fresh
by fountains; but I was most surprised to see the wal

Printing it this way, we can see that the sentences have some line break issues. This is normal and doesn't really mean much, but let's clean it either way.

In [50]:
letter_35 = letter_35.replace('\n', ' ')

In [51]:
sentences = sent_tokenize(letter_35)

In [52]:
counter = 1
for sentence in sentences:
    print(counter) 
    print(sentence)
    counter += 1

1
I HAVE had the advantage of very fine weather, all my journey; and as the summer is now in its beauty, I enjoyed the pleasure of fine prospects; and the meadows being full of all sorts of garden flowers, and sweet herbs, my berlin perfumed the air as it pressed them.
2
The grand signior furnished us with thirty covered waggons for our baggage, and five coaches of the country for my women.
3
We found the road full of the great spahis and their equipages coming out of Asia to the war.
4
They always travel with tents; but I chose to ly in houses all the way.
5
I will not trouble you with the names of the villages we passed, in which there was nothing remarkable, but at Ciorlei, where there was a _conac_, or little seraglio, built for the use of the grand signior, when he goes this road.
6
I had the curiosity to view all the apartments destined for the ladies of his court.
7
They were in the midst of a thick grove of trees, made fresh by fountains; but I was most surprised to see the wal

In [49]:
# When we split it sentence by sentence, we can see that the sentences have some line break issues
# Sometimes analyzing the text can reveal formatting issues that need to be addressed, 
# especially when dealing with historical texts or texts extracted from formats like ePub.

# Let's fix the line breaks in the actual letter

letter_35_test = letter_35.replace('\n', ' ')

letter_35_sentences_test = sent_tokenize(letter_35_test)

print(f"Number of sentences in letter 35: {len(letter_35_sentences_test)}")

counter = 1
for sentence in letter_35_sentences_test:
    print(counter) 
    print(sentence)
    counter += 1


Number of sentences in letter 35: 68
1
I HAVE had the advantage of very fine weather, all my journey; and as the summer is now in its beauty, I enjoyed the pleasure of fine prospects; and the meadows being full of all sorts of garden flowers, and sweet herbs, my berlin perfumed the air as it pressed them.
2
The grand signior furnished us with thirty covered waggons for our baggage, and five coaches of the country for my women.
3
We found the road full of the great spahis and their equipages coming out of Asia to the war.
4
They always travel with tents; but I chose to ly in houses all the way.
5
I will not trouble you with the names of the villages we passed, in which there was nothing remarkable, but at Ciorlei, where there was a _conac_, or little seraglio, built for the use of the grand signior, when he goes this road.
6
I had the curiosity to view all the apartments destined for the ladies of his court.
7
They were in the midst of a thick grove of trees, made fresh by fountains; bu

Much better for my eyes!

We will leave it at this for this notebook but in the next steps we will return to our sentences using the entire letter corpus!